org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ModelProviderUtil.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2018 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.buildship.core.internal.workspace;
+
+import java.util.Collection;
+import java.util.Set;
+
+import org.gradle.tooling.CancellationTokenSource;
+import org.gradle.tooling.model.eclipse.EclipseProject;
+
+import com.google.common.collect.ImmutableSet;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import org.eclipse.buildship.core.internal.util.gradle.HierarchicalElementUtils;
+
+public class ModelProviderUtil {
+
+    /**
+     * @return a flattened list of Eclipse projects for all projects included in the Gradle build
+     */
+    public static Set<EclipseProject> fetchAllEclipseProjects(GradleBuild build, CancellationTokenSource tokenSource, FetchStrategy fetchStrategy, IProgressMonitor monitor) {
```

Comment:
⭕️ I think we shouldn't use a different API than our users are using. If we think that caching is important, it should be part of the official API.

A possible solution: Decorate the `withConnection` method when invoked from project synchronization so that it automatically caches models that were requested during that synchronization before. This may help reduce the number of TAPI calls.

An even more efficient alternative would be an API where configurators can tell us what models they are interested in and we fetch them all in one go, including the ones we need.

This is not something we need to solve in this PR, but I wanted to point it out, because the current API encourages inefficiency (configuring the build many times by calling `withConnection` in different configurators).

Reply:
https://github.com/eclipse/buildship/issues/800